### DISCUSSION TITLE
Implementing Close Callback for Client Disconnect in Mixed Version Environment

### DISCUSSION THREAD

**Discussion by Fireshtorm1k** [zeroc-ice/ice#2738] [Q&A] [answered]
I am seeking assistance on how to implement a Close Callback when a client disconnects from the server. The challenge is that our client is developed using Ice version 3.7.10, while the server is running on Ice version 3.3.1. Unfortunately, we do not have access to the server's code.

Could you please advise on the best approach to achieve this functionality under these circumstances? If possible, providing an example similar to your Session example would be incredibly helpful.

Thank you for your time and assistance.

**Comment by pepone** [comment]
Hi,

The client close callback is independent of the server. It is just a matter of calling `setCloseCallback` on the connection to set your callback.

The C++ Glacier2/simpleChat uses this to detect if the session has been destroyed:

https://github.com/zeroc-ice/ice-demos/blob/77e7ba76ca31a6b67b306481b462d30a9cb6964a/cpp/Glacier2/simpleChat/Client.cpp#L102-L109

**Reply by Fireshtorm1k** [reply]
But how does the server know about closing the connection, provided that the client is suddenly disconnected?
How was it done correctly in older versions of Ice, where there was no setCloseCallback

**Reply by pepone** [reply, marked as answer]
In older versions, we used to have the client periodically call a server operation using a timer task, and then the server will cleanup the client if it didn't make any calls within the expected timeout.

See for example the 3.4 version of the Ice/session demo https://github.com/zeroc-ice/ice-demos/blob/3.4/demo/Ice/session/Client.cpp
